pkg/util/sync/pool.go

```diff
@@ -0,0 +1,42 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+// Package sync is utilities for synchronization
+package sync
+
+import "sync"
+
+// TypedPool is a type-safe version of sync.Pool
+type TypedPool[K any] struct {
+	p sync.Pool
+}
+
+// NewDefaultTypedPool creates a TypedPool using the default `new` function to create instances of K
+func NewDefaultTypedPool[K any]() *TypedPool[K] {
+	return NewTypedPool(func() *K {
+		return new(K)
+	})
+}
+
```

Comment:
Considering the number of use cases where the type is a slice, what do you think of adding a `NewSlicePool` constructor to make this case simpler ?

Eg. (disclaimer this is untested):
```go
func NewSlicePool[K any](size, capacity int) *TypedPool[[]K] {
	return NewTypedPool(func() *[]K {
		return &make([]K, size, capacity)
	})
}
```

There are 10 uses of `NewTypedPool`, and 8 of those are equivalent to this function.

Reply:
Added!